### ISSUE TITLE
`flake --exclude=.*` doesn't run any checks [REPLACEMENT ISSUE]

### ISSUE THREAD

**Issue by asottile** [PyCQA/flake8#215]
The original issue

	Id: 288
	Title: `flake --exclude=.*` doesn't run any checks

could not be created.
This is a dummy issue, replacing the original one. It contains everything but the original issue description. In case the gitlab repository is still existing, visit the following link to show the original issue:

TODO

**Comment by asottile** [comment]
In GitLab by @FichteFoll on Dec 27, 2016, 08:53

changed title from **`flake --excude=.*` doesn&#39;t run any checks** to **`flake --exc{+l+}ude=.*` doesn&#39;t run any checks**

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Dec 27, 2016, 17:01

closed

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Dec 27, 2016, 17:01

So let me explain why I'm not particularly bothered by this, since explaining it in IRC led to you trying to form an internet mob.

Checkout any version of pep8 or pycodestyle you want and run something like this:

```
python pycodestyle.py --exclude='.*' .
```

You get the exact same behaviour. This means that this has been the behaviour for quite a long time, so you've always used (at least) two tools that behave this way when provided with a rather naive filename glob. Before I told you on IRC that I didn't consider this a bug, I looked at the code myself. Here are some of the relevant bits that could be modified to "fix" this.

### src/flake8/utils.py

```py
def filenames_from(arg, predicate=None):                                                                                                
    # type: (str, callable) -> Generator                                                                                                
    """Generate filenames from an argument.                                                                                             
                                                                                                                                        
    :param str arg:                                                                                                                     
        Parameter from the command-line.                                                                                                
    :param callable predicate:                                                                                                          
        Predicate to use to filter out filenames. If the predicate                                                                      
        returns ``True`` we will exclude the filename, otherwise we                                                                     
        will yield it. By default, we include every filename                                                                            
        generated.                                                                                                                      
    :returns:                                                                                                                           
        Generator of paths                                                                                                              
    """                                                                                                                                 
    if predicate is None:                                                                                                               
        predicate = _default_predicate                                                                                                  
                                                                                                                                        
    if predicate(arg):                                                                                                                  
        return                                                                                                                          
                                                                                                                                        
    if os.path.isdir(arg):                                                                                                              
        for root, sub_directories, files in os.walk(arg):                                                                               
            if predicate(root):                                                                                                         
                sub_directories[:] = []                                                                                                 
                continue                                                                                                                
                                                                                                                                        
            # NOTE(sigmavirus24): os.walk() will skip a directory if you                                                                
            # remove it from the list of sub-directories.                                                                               
            for directory in sub_directories:                                                                                           
                joined = os.path.join(root, directory)                                                                                  
                if predicate(directory) or predicate(joined):                                                                           
                    sub_directories.remove(directory)                                                                                   
                                                                                                                                        
            for filename in files:                                                                                                      
                joined = os.path.join(root, filename)                                                                                   
                if predicate(joined) or predicate(filename):                                                                            
                    continue                                                                                                            
                yield joined                                                                                                            
    else:                                                                                                                               
        yield arg  
```

### src/flake8/checker.py

```py
class Manager(object):
    # ... [snip] ...
    def make_checkers(self, paths=None):                                                                                                
        # type: (List[str]) -> NoneType                                                                                                 
        """Create checkers for each file."""                                                                                            
        if paths is None:                                                                                                               
            paths = self.arguments                                                                                                      
                                                                                                                                        
        if not paths:                                                                                                                   
            paths = ['.']                                                                                                               
                                                                                                                                        
        filename_patterns = self.options.filename                                                                                       
        running_from_vcs = self.options._running_from_vcs                                                                               
                                                                                                                                        
        # NOTE(sigmavirus24): Yes this is a little unsightly, but it's our                                                              
        # best solution right now.                                                                                                      
        def should_create_file_checker(filename, argument):                                                                             
            """Determine if we should create a file checker."""                                                                         
            matches_filename_patterns = utils.fnmatch(                                                                                  
                filename, filename_patterns                                                                                             
            )                                                                                                                           
            is_stdin = filename == '-'                                                                                                  
            file_exists = os.path.exists(filename)                                                                                      
            # NOTE(sigmavirus24): If a user explicitly specifies something,                                                             
            # e.g, ``flake8 bin/script`` then we should run Flake8 against                                                              
            # that. Since should_create_file_checker looks to see if the                                                                
            # filename patterns match the filename, we want to skip that in                                                             
            # the event that the argument and the filename are identical.                                                               
            # If it was specified explicitly, the user intended for it to be                                                            
            # checked.                                                                                                                  
            explicitly_provided = (not running_from_vcs and                                                                             
                                   (argument == filename))                                                                              
            return ((file_exists and                                                                                                    
                     (explicitly_provided or matches_filename_patterns)) or                                                             
                    is_stdin)                                                                                                           
                                                                                                                                        
        checks = self.checks.to_dictionary()                                                                                            
        checkers = (                                                                                                                    
            FileChecker(filename, checks, self.options)                                                                                 
            for argument in paths                                                                                                       
            for filename in utils.filenames_from(argument,                                                                              
                                                 self.is_path_excluded)                                                                 
            if should_create_file_checker(filename, argument)                                                                           
        )                                                                                                                               
        self.checkers = [                                                                                                               
            checker for checker in checkers if checker.should_process                                                                   
        ]                                                                                                                               
        LOG.info('Checking %d files', len(self.checkers))  
```

## Option 1 - Hardcode a check for '.'

This could happen in `checker.py` where we either check for exclusion or if it's '.'. That's all kind of silly though.

## Option 2 - Update check in `filenames_from`

The section that reads

```py
if predicate(arg):
    return
```

Could be

```py
if predicate(arg) and not os.path.isdir(arg):
    return
```

But that would mean that we then have to begin to walk the directory before we actually exclude it. So I ran some tests using:

```py
import functools
import os.path

from flake8 import utils


def is_path_excluded(path, exclude):
    # type: (str) -> bool
    """Check if a path is excluded.

    :param str path:
        Path to check against the exclude patterns.
    :param list exclude:
        List of exclude patterns
    :returns:
        True if there are exclude patterns and the path matches,
        otherwise False.
    :rtype:
        bool
    """
    if not exclude:
        return False

    basename = os.path.basename(path)
    if utils.fnmatch(basename, exclude):
        return True

    absolute_path = os.path.abspath(path)
    match = utils.fnmatch(absolute_path, exclude)
    return match


def experiment(path, exclude):
    excluded = functools.partial(
        is_path_excluded,
        exclude=utils.normalize_paths(exclude),
    )
    number_of_results = len(list(utils.filenames_from(path, excluded)))
    print('Found {} looking in {!r} using {!r}'.format(
        number_of_results, '.', exclude))


def main():
    experiment('.', ['.*'])
    experiment('.', ['.*/'])
    experiment('.', ['flake8'])


if __name__ == '__main__':
    main()
```

Without that change

```
❯❯❯ tox -e venv -- python t.py
Found 386 looking in '.' using ['.*']
Found 386 looking in '.' using ['.*/']
Found 5592 looking in '.' using ['flake8']
```

With it

```
❯❯❯ tox -e venv -- python t.py
Found 0 looking in '.' using ['.*']
Found 386 looking in '.' using ['.*/']
Found 5594 looking in '.' using ['flake8']
```

So there's a slight difference, but not much in the case that you care about. (And no, I didn't bother trying to determine the 2 item difference, honestly.)

We also can't really modify the part inside the loop that checks `predicate(root)` because every `root` will be a directory. We could add a hardcoded check for `'.'`, but see `Option 1`.

## Option 3

Stick to what we've always had. It's simple. Yes. It's perhaps not the most intuitive, but it's better than trying to shove something in that will introduce an incompatibility between two previously compatible tools. (Yes both could be fixed, but frankly the code in pycodestyle is less well-factored and would require some refactoring to make it easier to "fix".)

Again, we could probably finagle some way of making this work, but I'm rather certain it will only introduce some unexpected behaviour at the same time.

I appreciate you taking the time to file this and provide your research, but I really think this will be too much work for too little value. Further, since you've insisted on making a big deal out of this, you've already caused me to spend more time on this than I wanted to, so even if you manage to incite a mob, I'm going to happily ignore this.

**Comment by asottile** [comment]
In GitLab by @FichteFoll on Dec 27, 2016, 17:56

**TL;DR**: "it's not a bug" != "it's a bug, but I won't fix it"

---

Thank you for you time with explaining this in detail to me. It didn't have to be this long (since I made myself familiar with the relevant codebase already), but it is appreciated.

>So let me explain why I'm not particularly bothered by this, since explaining it in IRC led to you trying to form an internet mob.

There seems to be a misunderstanding here. First, you didn't really explain anything. I'll recap our small discussion in its entirety:

```
[02:50:15]  --> You (FichteFoll!~Fichte@unaffiliated/fichtefoll) joined ##python-code-quality
[02:50:37]  <FichteFoll> any idea why "flake8 ." doesn't scan any files but "flake8 subdir" does?
[02:51:25]  <FichteFoll> my tox.ini has this: https://6.lithi.io/6lhAz.txt
[02:52:18]  <FichteFoll> okay, so it appears it's the `exclude = .*` line...
[02:52:43]  <FichteFoll> this might even be a bug
[15:04:10]  <sigmavirus24> FichteFoll: if it's a bug, it's a bug in the standard library
[18:52:28]  <FichteFoll> what are you using to implement the exclude filters, sigmavirus24?
[19:59:10]  <sigmavirus24> os.path.expandpath (if there's a / in the exclude glob) and glob.glob iirc
[19:59:36]  <sigmavirus24> or fnmatch, can't remember which
[20:02:51]  <FichteFoll> either way, can you comment on this being expected or unexpected behavior?
[20:22:05]  <sigmavirus24> FichteFoll: I'm frankly not surprised by it, but I'm willing to bet that's always been the behaviour
[20:22:26]  <sigmavirus24> Either way, I think .* is a bit too silly to bother special-casing the code in flake8 right now
[20:22:32]  <sigmavirus24> I'd rather you just do ./.*
[20:22:58]  <sigmavirus24> That'll do what you want and that's what I've had in one of my projects for a few years now
[20:24:53]  <FichteFoll> hm yeah, that could work too
[22:43:46]  <FichteFoll> as a matter of fact, falke8 uses fnmatch
[22:47:30]  <FichteFoll> flake8.checker            MainProcess    342 DEBUG    "." has been excluded
[22:47:34]  <FichteFoll> why does it even check for that
[22:51:50]  <FichteFoll> okay, it appears that it goes like this:
[22:53:58]  <FichteFoll> 1. path arguments to flake8 (by default ".") are passed to utils.filenames_from, which is supposed to yield all absolute (!) filepaths from a root
[22:55:02]  <FichteFoll> 2. however, the predicate (checker.Manager.is_file_excluded) is executed on the argument *before* expanding it to an absolute path
[22:55:38]  <FichteFoll> while later actually running the predicate on the root yielded by os.walk again (where it is an absolute path)
[22:56:08]  <FichteFoll> 3. so, because the "." argument matches the ".*" pattern via fnmatch.fnmatch, it is excluded and no checkers are created at all
[22:56:36]  <FichteFoll> if you agree that this is a bug, I'll try working on a fix
[13:52:10]  <sigmavirus24> FichteFoll: I don't agree that it's a bug.
```

I read this as "you don't think it's a bug" instead of "it's probably a bug, but fixing it wouldn't be feasable", because that's literally what you ~~said~~ wrote. With this understanding, I went to create a proper report of what I believed to be a bug and asked others for their opinion since we were apparently mutually disagreeing. It was definitely *not* my intention to form an "internet mob"; I merely wanted to gather others' opinions to wage my own. From your wording it appears that you were directly approached by others regarding this matter, which was certainly not what I wanted. I apologize if this made you uncomfortable and in hindsight I would not have mentioned that you already commented on the report when I mentioned this in the #positivepython channel.

I also don't think it's unlikely that this behavior has existed for a long time and went mostly unnoticed, but in my opinion this should not prevent bug fixes at all. Even more so, having an archivable report of something that could considered to be a bug but was not going to be fixed is valuable for any sort of project.

---

Now, onto the actual issue, I generally agree with your options, but there is another one (which gets more complicated the more you think about it).

## Option 4

In [`checker.py`](https://gitlab.com/pycqa/flake8/blob/master/src/flake8/checker.py#L209), run `os.path.abspath` on each argument from the command line before doing anything with it. This will magically unwrap the `.` current folder to the actual folder and do the same for all other relative paths. Absolute paths would be unaffected. For the entire following code, only absolute paths matter anyway. `fnmatch.fnmatch` is run both the entire path and its basename.

Afaics, there is only one situation where this would produce different results (besides `.*` ofc), which is when an exclude pattern would exclude the current directory's name, which should probably be taken care of specifically. However, there is more to this, which depends on how configuration is parsed from files and whether or not it should be possible to exclude the "current directory". Frankly, I don't know enough about the internals of flake8 I don't want to get too involved with it.

Disclaimer: I haven't yet understood why `./.*` works in the first place, so I might be missing something here.

That said, not fixing this edge case is a perfectly fine decision that I can relate to, considering how many internals and philosophies it actually affects.

**Comment by asottile** [comment]
In GitLab by @sigmavirus24 on Dec 28, 2016, 06:30

> Now, onto the actual issue, I generally agree with your options, but there is another one **(which gets more complicated the more you think about it).** [emphasis mine]

Right, I discarded that option entirely because while Flake8 3.0 broke a lot of things to make people generally happier, I'm not going to break more things in a minor or patch release. 

Creating absolute paths first breaks the existing behaviour pretty severely. We already check the absolute path, but we do so *after* checking the relative path. And yes, we do need to check both relative and absolute. People expect both behaviours and have come to expect that for the entire lifetime of Flake8's existence (or at least for as long as it relied on pep8/pycodestyle to do option handling, exclude handling, etc.). In fact if you look at `filenames_from` you'll notice this:

```py
            # NOTE(sigmavirus24): os.walk() will skip a directory if you                                                                
            # remove it from the list of sub-directories.                                                                               
            for directory in sub_directories:                                                                                           
                joined = os.path.join(root, directory)                                                                                  
                if predicate(directory) or predicate(joined):                                                                           
                    sub_directories.remove(directory)                                                                                   

            for filename in files:                                                                                                      
                joined = os.path.join(root, filename)                                                                                   
                if predicate(joined) or predicate(filename):                                                                            
                    continue                                                                                                            
                yield joined  
```

In the first for-loop we check the directory name itself, e.g., if we're traversing Flake8's source and `root` is `src` then `directory` might be `flake8`. So first we check if `flake8` matches an exclude pattern. If it doesn't then we check if `src/flake8` matches one. I can't remember right now if `os.walk` returns absolute or relative root names (I'm 90% certain it's relative), so we could probably expand that, but if the relative directory name is excluded then I'm not sure we will ever quite get there.

> I haven't yet understood why ./.* works in the first place, so I might be missing something here.

So let's look at the test file I was running experiments with, specifically these lines in the `experiment` function:

```py
    excluded = functools.partial(
        is_path_excluded,
        exclude=utils.normalize_paths(exclude),
    )
```

We use `utils.normalize_paths` from Flake8 because that's what's run against the `--exclude` option. It does two things:

1. It ensures that what it returns is a list of strings
1. If the strings have a `/` in them (or really, either of `os.path.sep` or `os.path.altsep`) it will call `os.path.abspath` on the pattern. So `./.*` becomes `/path/to/directory/.*`.

Then when you do `fnmatch.fnmatch(filename, '/path/to/directory/.*')` you will only get a match for a filename that looks like `/path/to/directory/.git` or `/path/to/directory/.tox`. Make sense?

> There seems to be a misunderstanding here. First, you didn't really explain anything.

I did, but it looks like either your client dropped it or mine did.
